scrapper.py

```diff
@@ -31,67 +53,226 @@ class Crawler:
-        pass
+        return datetime.fromisoformat(date_str)
 
-    def parse(self):
+    def parse(self) -> None:
         """
         Parses each article
         """
-        pass
+        try:
+            response = requests.get(self.full_url, headers=HEADERS, cookies=COOKIES)
+            response.raise_for_status()
+        except RequestException as exc:
+            log.exception("%s was encountered while parsing the article", exc)
+            return
+
+        soup = BeautifulSoup(response.text, features='lxml')
+
+        self._fill_article_with_text(soup)
+        self._fill_article_with_meta_information(soup)
 
+        self.article.save_raw()
 
-def prepare_environment(base_path):
+
+def prepare_environment(base_path: str) -> None:
     """
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    shutil.rmtree(base_path, ignore_errors=True)
+    Path(base_path).mkdir(parents=True, exist_ok=True)
 
 
-def validate_config(crawler_path):
+def validate_config(crawler_path: str) -> Iterable:
     """
     Validates given config
     """
-    pass
+    with open(crawler_path) as file:
+        config: dict = json.load(file)
+
+    Check = namedtuple('Check', ['status', 'error'])
+
+    try:
+        total_num = config['total_articles_to_find_and_parse']
+
+        is_correct_total_num = isinstance(total_num, int)
+        is_correct_url = all(isinstance(x, str) for x in config['base_urls'])
+        is_num_not_oor = is_correct_total_num and 0 < total_num <= 100000
+
+        checks = (
+            Check(is_correct_url, IncorrectURLError),
+            Check(is_correct_total_num, IncorrectNumberOfArticlesError),
+            Check(is_num_not_oor, NumberOfArticlesOutOfRangeError),
+        )
+
+        for check in checks:
+            if not check.status:
+                raise check.error
+
+        return config.values()
+
+    except (JSONDecodeError, KeyError) as exc:
+        log.exception("%s was encountered while validating crawler config.", exc)
+        raise UnknownConfigError from exc
 
 
 if __name__ == '__main__':
-    # YOUR CODE HERE
-    pass
+    urls, max_num_articles, max_per_seed = validate_config(CRAWLER_CONFIG_PATH)
+    crawler = CrawlerRecursive(seed_urls=urls, max_articles=max_num_articles, max_articles_per_seed=max_per_seed)
+
+    if not os.path.exists(ASSETS_PATH):
+        prepare_environment(ASSETS_PATH)
+
+    crawler.find_articles()
+    seed, articles, parsed = crawler.load_interim()
+
+    for idx, article_url in enumerate(sorted(crawler.urls)):
+        parser = ArticleParser(full_url=article_url, article_id=idx)
+        time.sleep(random.uniform(2, 4))
+
+        if article_url in parsed:
+            log.info("Article #%s '%s' is already processed. Skipping.", idx, article_url)
+            continue
+
+        parser.parse()
+        parsed.append(article_url)
+
+        crawler.save_interim(seed=seed, articles=list(articles), processed=parsed)
```

Comment:
crawler knows nothing about processed files

Reply:
addressed this issue